trusted-content: restructure doi section (#19745)
docker/docs: content/trusted-content/official-images/using.md

---
title: Using Docker Official Images
description: |
  Learn about building applications with Docker Official images
  and how to interpret the tag names they use.
keywords: docker official images, doi, tags, slim, feedback, troubleshooting
weight: 10
---

Docker recommends you use the Docker Official Images in your projects.
These images have clear documentation, promote best practices, and are regularly updated.
Docker Official Images support most common use cases, making them perfect for new Docker users.
Advanced users can benefit from more specialized image variants as well as review Docker Official Images as part of your `Dockerfile` learning process.

## Tags

The repository description for each Docker Official Image contains a
**Supported tags and respective Dockerfile links** section that lists all the
current tags with links to the Dockerfiles that created the image with those
tags. The purpose of this section is to show what image variants are available.

> ![Example: supported tags for Ubuntu](../images/supported_tags.webp)

Reviewer: Given this is a screenshot of plain markdown as rendered by Hub, would it be cleaner to just embed that same markdown in a quoted section here instead of adding a binary blob of an image (and in a file format that GitHub doesn't natively even recognize as an image, so it doesn't show up in diffs)?  That would make it easier to update over time if you wanted or needed to:

```suggestion
> # Supported tags and respective `Dockerfile` links
>
> -	[`20.04`, `focal-20240216`, `focal`](https://git.launchpad.net/cloud-images/+oci/ubuntu-base/tree/oci/index.json?h=refs/tags/dist-focal-amd64-20240216-5e82aa8e&id=5e82aa8e3498567df6757d3e9598a8cd54d19ee9)
> -	[`22.04`, `jammy-20240227`, `jammy`, `latest`](https://git.launchpad.net/cloud-images/+oci/ubuntu-base/tree/oci/index.json?h=refs/tags/dist-jammy-amd64-20240227-c988e436&id=c988e4364bc3cc0378f6c3592eb486d281a6ac41)
> -	[`23.10`, `mantic-20240216`, `mantic`, `rolling`](https://git.launchpad.net/cloud-images/+oci/ubuntu-base/tree/oci/index.json?h=refs/tags/dist-mantic-amd64-20240216-c9ad3939&id=c9ad39394a666b231d5c5398e3cd38ac70c0594c)
> -	[`24.04`, `noble-20240225`, `noble`, `devel`](https://git.launchpad.net/cloud-images/+oci/ubuntu-base/tree/oci/index.json?h=refs/tags/dist-noble-amd64-20240225-8cf671e3&id=8cf671e3c71fa7d0735ceefd4a77c7af2771472f)
```

Author: I think the screenshot adds an extra touch of familiarity if you've come across that readme section on Docker Hub, but not quite understood it. I'd prefer to keep the screenshot for that reason I think

Tags listed on the same line all refer to the same underlying image. Multiple
tags can point to the same image. For example, in the previous screenshot taken
from the `ubuntu` Docker Official Images repository, the tags `24.04`,
`noble-20240225`, `noble`, and `devel` all refer to the same image.

The `latest` tag for a Docker Official Image is often optimized for ease of use
and includes a wide variety of useful software, such as developer and build tools.
Because of their ease of use and wide applicability, `latest` images are often

> used in getting-started guides, but they're typically not optimized for production use.

Reviewer: The way I usually explain to maintainers for how to choose which variant to put `latest` on is "Which image would you suggest users should use if they have no idea or are unfamiliar with the underlying software?" (which I think more clearly highlights why `ubuntu:latest` is the latest LTS, not the latest release :eyes:)

## Slim images

A number of operating system and language stacks such as
[Debian](https://hub.docker.com/_/debian/),
[Node.js](https://hub.docker.com/_/node/),
[Python](https://hub.docker.com/_/python/), and
[Ruby](https://hub.docker.com/_/ruby/) have `slim` tag variants
designed to provide a lightweight, production-ready base image

> with fewer packages.

Reviewer: I think we should probably remove Debian from this list because the slim variants of Debian are not quite the same (even if they're in the same spirit) as these others -- they have the same packages, but less of the installed files from those packages (most notably, they exclude documentation not just from the packages in the image, but from all packages installed afterwards).

Reviewer 2: Should we also remove "operating system and" from the lead in, since Debian and Oracle Linux are the only OS base image repos that offer slim variants, and both are similarly not recommended?

A typical consumption pattern for `slim`
images is as the base image for the final stage of a
[multi-staged build](https://docs.docker.com/build/building/multi-stage/).
For example, you build your application in the first stage of the build
using the `latest` variant and then copy your application into the final
stage based upon the `slim` variant. Here is an example `Dockerfile`.

```dockerfile
FROM node:latest AS build
WORKDIR /app
COPY package.json package-lock.json ./
RUN npm ci
COPY . ./
FROM node:slim
WORKDIR /app
COPY --from=build /app /app
CMD ["node", "app.js"]
```

> ## Alpine

Reviewer: Should these following subheadings have the "images" suffix like "Slim images" does above, or should it not have the suffix?  (It feels like for consistency, we should probably do one or the other)

Author: sure, let me add `images`

Many Docker Official Images repositories also offer `alpine` variants. These
images are built on top of the [Alpine Linux](https://www.alpinelinux.org/)
distribution rather than Debian or Ubuntu. Alpine Linux is focused on providing
a small, simple, and secure base for container images, and Docker Official
Images `alpine` variants typically aim to install only necessary packages. As a
result, Docker Official Images `alpine` variants are typically even smaller
than `slim` variants.

It's worth keeping in mind that Alpine Linux is based on musl libc, as opposed
to glibc, which is used by most other Linux distributions. This difference in C
implementations may influence how your program runs, depending on the type of
application you're building. With Alpine Linux, your program can't use
dynamically linked glibc objects, as those aren't available on musl. To run
your programs on Alpine Linux, you can:

- Compile your program against musl libc
- Statically link glibc libraries into your program
- Avoid C dependencies altogether (for example, build Go programs without CGO)

> - Install a glibc compatibility layer manually, alongside musl

Reviewer: We should definitely without a doubt drop this last bullet point (see references linked from https://github.com/docker-library/official-images/issues/10794).

I also wonder if we should work in more of the content from our standard "Alpine blurb" here? https://github.com/docker-library/docs/blob/c5a57f3ec8904a2232986c1bd18f45558482eee1/.template-helpers/variant-alpine.md (maybe one of the links?  maybe a reference to where to find this information on Hub, because it sometimes changes per-image, with `golang:alpine` being a notable example: https://github.com/docker-library/docs/blob/c5a57f3ec8904a2232986c1bd18f45558482eee1/golang/variant-alpine.md)

Author: Included some of the wordings from there. I prefer not to link out to the orange website or issues on other projects if I can help it.

These images have clear documentation, promote best practices, and are regularly updated.
Docker Official Images support most common use cases, making them perfect for new Docker users.
Advanced users can benefit from more specialized image variants as well as review Docker Official Images as part of your `Dockerfile` learning process.

## Tags

The repository description for each Docker Official Image contains a
**Supported tags and respective Dockerfile links** section that lists all the
current tags with links to the Dockerfiles that created the image with those
tags. The purpose of this section is to show what image variants are available.

![Example: supported tags for Ubuntu](../images/supported_tags.webp)

Tags listed on the same line all refer to the same underlying image. Multiple
tags can point to the same image. For example, in the previous screenshot taken
from the `ubuntu` Docker Official Images repository, the tags `24.04`,
`noble-20240225`, `noble`, and `devel` all refer to the same image.

The `latest` tag for a Docker Official Image is often optimized for ease of use
and includes a wide variety of useful software, such as developer and build tools.

> Because of their ease of use and wide applicability, `latest` images are often
> used in getting-started guides, but they're typically not optimized for production use.

Reviewer 2: ```suggestion
By tagging an image as `latest`, the image maintainers are essentially suggesting
that image be used as the default. In other words, if you do not know what tag to
use or are unfamiliar with the underlying software, you should probably start with
the `latest` image. As your understanding of the software and image variants advances,
you may find other image variants better suit your needs.
```

How's this @tianon-sso ?

Reviewer: This sounds like you've taken my _own_ words and adapted them to fit! :joy: :heart:

(LGTM)

Reviewer 2: I am an editor, not a writer

designed to provide a lightweight, production-ready base image
with fewer packages.

A typical consumption pattern for `slim`
images is as the base image for the final stage of a
[multi-staged build](https://docs.docker.com/build/building/multi-stage/).
For example, you build your application in the first stage of the build
using the `latest` variant and then copy your application into the final
stage based upon the `slim` variant. Here is an example `Dockerfile`.

```dockerfile
FROM node:latest AS build
WORKDIR /app
COPY package.json package-lock.json ./
RUN npm ci
COPY . ./
FROM node:slim
WORKDIR /app
COPY --from=build /app /app
CMD ["node", "app.js"]
```

## Alpine images

Many Docker Official Images repositories also offer `alpine` variants. These
images are built on top of the [Alpine Linux](https://www.alpinelinux.org/)
distribution rather than Debian or Ubuntu. Alpine Linux is focused on providing
a small, simple, and secure base for container images, and Docker Official
Images `alpine` variants typically aim to install only necessary packages. As a
result, Docker Official Images `alpine` variants are typically even smaller
than `slim` variants.

> The main caveat to note is that it uses [musl libc](https://musl.libc.org/)

Reviewer 2: ```suggestion
The main caveat to note is that Alpine Linux uses [musl libc](https://musl.libc.org/)
```

instead of [glibc](https://www.gnu.org/software/libc/). Additionally, to
minimize image size, it's uncommon for Alpine-based images to include tools
such as Git or Bash by default. Depending on the depth of libc requirements or
assumptions in your programs, you may find yourself running into issues due to
missing libraries or tools.

When you use Alpine images as a base, consider the following options in order
to make your program compatible with Alpine Linux and musl:

- Compile your program against musl libc
- Statically link glibc libraries into your program
- Avoid C dependencies altogether (for example, build Go programs without CGO)

> - Add the things you need in your own Dockerfile.

Reviewer 2: ```suggestion
- Add the software you need yourself in your Dockerfile.
```

Refer to the `alpine` image [description](https://hub.docker.com/_/alpine) on
Docker Hub for examples on how to install packages if you are unfamiliar.

## Codenames

Tags with words that look like Toy Story characters (for example, `bookworm`,
`bullseye`, and `trixie`) or adjectives (such as `focal`, `jammy`, and
`noble`), indicate the codename of the Linux distribution they use as a base
image. Debian release codenames are [based on Toy Story characters](https://en.wikipedia.org/wiki/Debian_version_history#Naming_convention),
and Ubuntu's take the form of "Adjective Animal". For example, the
codename for Ubuntu 24.04 is "Noble Numbat".

Linux distribution indicators are helpful because many Docker Official Images
provide variants built upon multiple underlying distribution versions (for
example, `postgres:bookworm` and `postgres:bullseye`).

> 

Reviewer 2: ```suggestion

## Other tags
```

a complete build toolchain to support installation of modules that depend on
optimized code. An advanced user could build a custom image with just the
necessary pre-compiled libraries to save space.

## Tags

The repository description for each Docker Official Image contains a
**Supported tags and respective Dockerfile links** section that lists all the
current tags with links to the Dockerfiles that created the image with those
tags. The purpose of this section is to show what image variants are available.

![Supported tags for Ubuntu](../images/supported_tags.webp)

Tags listed on the same line all refer to the same underlying image. Multiple
tags can point to the same image. For example, in the `ubuntu` Docker Official
Images repository, the tags `24.04`, `noble-20240225`, `noble`, and `devel` all
refer to the same image.

The `latest` tag for a Docker Official Image is often optimized for ease of use
and includes a wide variety of useful software, such as Git and build tools.
Because of their ease of use and wide applicability, `latest` images are often
used in getting-started guides, but they're not recommended for production use.

> ## Slim images

Author: Yes that would be good. We also have this page which is related, but a little bit of an oddity in the "best practices" section: https://docs.docker.com/build/building/base-images/

Not sure what to do about that if anything. Just an observation... Maybe there's a home for "base images" somewhere.

> ## Slim images

Reviewer 3: Oh, interesting; I recall it used to be debian (as a default?), but maybe that changed (debian generally considered more "stable", but ubuntu having more up to date packages)

> ## Slim images

Reviewer: If we look at repositories instead of individual tags, we get a slightly different story (the point about Ubuntu still holds, but it also helps highlight that Eclipse Temurin is a special case because Java is, not because Temurin itself really is):

```console
$ bashbrew children --uniq debian | cut -d: -f1 | sort -u | wc -l
66
$ bashbrew children --uniq ubuntu | cut -d: -f1 | sort -u | wc -l
45
$ bashbrew children --uniq eclipse-temurin | cut -d: -f1 | sort -u | wc -l
23
```

implementations may influence how your program runs, depending on the type of
application you're building. With Alpine Linux, your program can't use
dynamically linked glibc objects, as those aren't available on musl. To run
your programs on Alpine Linux, you can:

- Compile your program against musl libc
- Statically link glibc libraries into your program
- Avoid C dependencies altogether (for example, build Go programs without CGO)
- Install a glibc compatibility layer manually, alongside musl

## Codenames

Tags with words that look like Toy Story characters (for example, `bookworm`,
`bullseye`, and `trixie`) or adjectives (such as `focal`, `jammy`, and
`noble`), indicate the codename of the Linux distribution they use as a base
image. Debian-release codenames are [based on Toy Story characters](https://en.wikipedia.org/wiki/Debian_version_history#Naming_convention),
and Ubuntu's take the form of "Adjective Animal". For example, the
codename for Ubuntu 24.04 is "Noble Numbat".

Linux distribution indicators are helpful because many Docker Official Images
provide variants built upon multiple underlying distribution versions (for

> example, `postgres:bookworm` and `postgres:bullseye`).

Reviewer 2: Can something like this be added to the end?

```
Docker Official Images tags may contain other hints to the purpose of
their image variant in addition to those described above. Often these
tag variants are explained later in the Docker Official Images repository
documentation. Reading through the “How to use this image” and/or
“Image Variants” sections will help you to understand how to use these
variants.
```
